Address performance regression introduced by #90218

JakobDegen opened the pull request:
As part of the changes in #90218 , the `adt_drop_tys` and friends code stopped recursing through the query system, meaning that intermediate computations did not get cached. This change adds the recursions back in without re-introducing any of the old issues.

On local benchmarks this fixes the 5% regressions in #90504 ; the wg-grammar regressions didn't seem to move too much. I may take some time later to look into those.

Not sure who to request for review here, so will leave it up to whoever gets it.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

matthewjasper commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit e0269a0aa302126ff020ff09791bfde05f8620dd with merge 5edbd592d7d24d3910283d67f944fc057d860f7d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e0269a0aa302126ff020ff09791bfde05f8620dd","merge_sha":"5edbd592d7d24d3910283d67f944fc057d860f7d"} -->

Mark-Simulacrum commented:
(Feel free to review as well, but I can try to help if you don't have a chance)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4196525916?check_suite_focus=true)
Build commit: 5edbd592d7d24d3910283d67f944fc057d860f7d (`5edbd592d7d24d3910283d67f944fc057d860f7d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4196525916?check_suite_focus=true"},"merge_sha":"5edbd592d7d24d3910283d67f944fc057d860f7d"} -->

rust-timer commented:
Queued 5edbd592d7d24d3910283d67f944fc057d860f7d with parent e90c5fbbc5df5c81267747daeb937d4e955ce6ad, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e90c5fbbc5df5c81267747daeb937d4e955ce6ad&end=5edbd592d7d24d3910283d67f944fc057d860f7d).

rust-timer commented:
Finished benchmarking commit (5edbd592d7d24d3910283d67f944fc057d860f7d): [comparison url](https://perf.rust-lang.org/compare.html?start=e90c5fbbc5df5c81267747daeb937d4e955ce6ad&end=5edbd592d7d24d3910283d67f944fc057d860f7d).

**Summary**: This change led to very large relevant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -5.1% on `full` builds of `regression-31157`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

the8472 commented on compiler/rustc_ty_utils/src/needs_drop.rs:
```diff
@@ -203,22 +210,46 @@ fn drop_tys_helper<'tcx>(
+                    }
+                    _ => vec.push(subty),
+                };
+                Ok(vec)
```
Imo a `for` loop would be more readable here.

the8472 commented on compiler/rustc_ty_utils/src/needs_drop.rs:
```diff
@@ -186,11 +192,12 @@ fn drop_tys_helper<'tcx>(
+        let mut subreqs = if adt_def.is_manually_drop() {
             debug!("drop_tys_helper: `{:?}` is manually drop", adt_def);
-            return Ok(Vec::new().into_iter());
+            Vec::new().into_iter()
```
All those `into_iter` could be hoisted out of the if/else blocks

JakobDegen commented:
Refactored more generally and extracted some of the uglyness out into a function, this both avoids the double allocation and makes things more readable

Mark-Simulacrum commented:
@bors try @rust-timer queue

but I think this looks good to me -- presuming perf is still good after refactoring (as expected), r=me

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 746091c61068067695ba0f910651a4b4d97d10b7 with merge 922d360035cc6fd56a30063435326ce1495ea99c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"746091c61068067695ba0f910651a4b4d97d10b7","merge_sha":"922d360035cc6fd56a30063435326ce1495ea99c"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4204575471?check_suite_focus=true)
Build commit: 922d360035cc6fd56a30063435326ce1495ea99c (`922d360035cc6fd56a30063435326ce1495ea99c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4204575471?check_suite_focus=true"},"merge_sha":"922d360035cc6fd56a30063435326ce1495ea99c"} -->

rust-timer commented:
Queued 922d360035cc6fd56a30063435326ce1495ea99c with parent 3b2c45441d7eefed63f6658ff8becd5a51eaeae1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3b2c45441d7eefed63f6658ff8becd5a51eaeae1&end=922d360035cc6fd56a30063435326ce1495ea99c).

rust-timer commented:
Finished benchmarking commit (922d360035cc6fd56a30063435326ce1495ea99c): [comparison url](https://perf.rust-lang.org/compare.html?start=3b2c45441d7eefed63f6658ff8becd5a51eaeae1&end=922d360035cc6fd56a30063435326ce1495ea99c).

**Summary**: This change led to very large relevant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -5.1% on `full` builds of `regression-31157`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

JakobDegen commented:
@Mark-Simulacrum perf seems to look good, but I believe I don't have perms to r= you

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 746091c61068067695ba0f910651a4b4d97d10b7 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 746091c61068067695ba0f910651a4b4d97d10b7 -->
<!-- homu: {"type":"Approved","sha":"746091c61068067695ba0f910651a4b4d97d10b7","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 746091c61068067695ba0f910651a4b4d97d10b7 with merge 3a744751d561e69b5e3e5468db5223ad04c81368...
<!-- homu: {"type":"BuildStarted","head_sha":"746091c61068067695ba0f910651a4b4d97d10b7","merge_sha":"3a744751d561e69b5e3e5468db5223ad04c81368"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4206951909?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/3a744751d561e69b5e3e5468db5223ad04c81368/checks/4206951909/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

JakobDegen commented:
Is this a genuine failure or spurious? I'm not terribly great at reading Bors logs

the8472 commented:
Segfault during bootstrap, CI has been having a few of those lately (#90812)

@bors retry

bors commented:
:hourglass: Testing commit 746091c61068067695ba0f910651a4b4d97d10b7 with merge a2a7683e8f321e3c0b9d648ae480827b5ab70e1d...
<!-- homu: {"type":"BuildStarted","head_sha":"746091c61068067695ba0f910651a4b4d97d10b7","merge_sha":"a2a7683e8f321e3c0b9d648ae480827b5ab70e1d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4222110650?check_suite_focus=true)
Approved by: Mark-Simulacrum
Pushing a2a7683e8f321e3c0b9d648ae480827b5ab70e1d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4222110650?check_suite_focus=true"},"merge_sha":"a2a7683e8f321e3c0b9d648ae480827b5ab70e1d"} -->

rust-timer commented:
Finished benchmarking commit (a2a7683e8f321e3c0b9d648ae480827b5ab70e1d): [comparison url](https://perf.rust-lang.org/compare.html?start=02063124f96aac10f8c5c70653242b0704d397e0&end=a2a7683e8f321e3c0b9d648ae480827b5ab70e1d).

**Summary**: This change led to very large relevant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -5.1% on `full` builds of `regression-31157`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
